Common/Orders/Fills/EquityFillModel.cs

```diff
@@ -493,12 +495,58 @@ public virtual OrderEvent MarketOnOpenFill(Security asset, MarketOnOpenOrder ord
             var localOrderTime = order.Time.ConvertFromUtc(asset.Exchange.TimeZone);
-            if (currentBar == null || localOrderTime >= currentBar.EndTime) return fill;
+            var endTime = DateTime.MinValue;
+
+            var subscribedTypes = GetSubscribedTypes(asset);
+
+            if (subscribedTypes.Contains(typeof(Tick)))
+            {
+                var primaryExchange = (byte) ((Equity) asset).PrimaryExchange;
+                var officialOpen = (uint) (TradeConditionFlags.Regular | TradeConditionFlags.OfficialOpen);
+                var openingPrints = (uint) (TradeConditionFlags.Regular | TradeConditionFlags.OpeningPrints);
+
+                // Get the first valid (non-zero) tick of trade type from an open market
+                var trade = asset.Cache.GetAll<Tick>()
+                    .Where(x => !string.IsNullOrWhiteSpace(x.SaleCondition))
+                    .FirstOrDefault(x =>
+                        x.TickType == TickType.Trade && x.Price > 0 && x.ExchangeCode == primaryExchange &&
+                        (x.ParsedSaleCondition == officialOpen || x.ParsedSaleCondition == openingPrints) &&
+                        asset.Exchange.DateTimeIsOpen(x.Time));
+
+                if (trade != null)
+                {
+                    endTime = trade.EndTime;
+                    fill.FillPrice = trade.Price;
+                }
+            }
+            else if (subscribedTypes.Contains(typeof(TradeBar)))
+            {
+                var tradeBar = asset.Cache.GetData<TradeBar>();
+                if (tradeBar != null)
+                {
+                    // If the order was placed during the bar aggregation, we cannot use its open price
+                    if (tradeBar.Time < localOrderTime) return fill;
+
+                    // If the first data after the market opens is a quote bar (high resolution only),
+                    // the open of the last trade bar is not the open for the current day, but the last open in cache.
+                    // We need to verify whether the trade data is from the open market.
+                    if (tradeBar.Period < Resolution.Hour.ToTimeSpan() && !asset.Exchange.DateTimeIsOpen(tradeBar.Time))
+                    {
+                        return fill;
+                    }
+
+                    endTime = tradeBar.EndTime;
+                    fill.FillPrice = tradeBar.Open;
+                }
+            }
+
+            if (localOrderTime >= endTime) return fill;
 
             // if the MOO was submitted during market the previous day, wait for a day to turn over
-            if (asset.Exchange.DateTimeIsOpen(localOrderTime) && localOrderTime.Date == asset.LocalTime.Date)
+            // The date of the order and the trade data end time cannot be the same.
+            // Note that the security local time can be ahead of the data end time.
+            if (asset.Exchange.DateTimeIsOpen(localOrderTime) && localOrderTime.Date == endTime.Date)
```

Comment:
Not sure this is right always? see from IB https://www.interactivebrokers.eu/en/index.php?f=41323
![image](https://user-images.githubusercontent.com/18473240/123721418-b3333300-d85c-11eb-8106-94ec2b15c690.png)

Reply:
This PR will accept orders that are submitted until 9:30 (regular market closed).
See unit tests [here](https://github.com/QuantConnect/Lean/pull/5679/files#diff-18938dbae43240eab6eee79fdd2221594549c57b2b4937ba737188dd5197d9d5R690)
I think we could add the extra 2-minute, but I guess it's an edge case for Nasdaq.

Reply:
Your right! Ignore me 👍